src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -258,24 +254,22 @@ class InstanceOpFactoryImpl(
         // All of these cases are handled in one way: by creating a new taskId for a resident task based on the previous
         // one. The used function will increment the attempt counter if it exists, of append a 1 to denote the first attempt
         // in version 1.5.
-        val newTaskId = Task.Id.forResidentTask(taskId)
+        val taskIds: Seq[Task.Id] = {
+          val originalIds = if (reservedInstance.tasksMap.nonEmpty) {
+            reservedInstance.tasksMap.keys
+          } else {
+            Seq(Task.Id.forInstanceId(reservedInstance.instanceId, None))
+          }
+          originalIds.map(ti => Task.Id.forResidentTask(ti)).to[Seq]
+        }
+        val newTaskId = taskIds.headOption.getOrElse(throw new IllegalStateException(s"Expecting to have a task id present when creating instance for app ${app.id} from instance $reservedInstance"))
```

Comment:
the if is necessary because tasksMap is intentionally empty for scheduled instances. Now we have 2 cases:
- newly scheduled instance (that was not previously launched) - this one does not have dummy task anymore - so since we did not create that dummy Id when scheduling, we have to create it here
- we are rescheduling instance what was already terminal. In this case we keep the id around so that we can increment attempt (in general we should move away from this in the future and keep just the attempts count somewhere, not the whole task)

if you know how to express this better, let me know